File: client/src/network/protocol.js
// Message type definitions
export const MessageType = {
  // Client -> Server
  JOIN: 'join',
  AUDIO: 'audio',
  UTTERANCE_END: 'utterance_end',
  LEAVE: 'leave',
  PING: 'ping',

  // Server -> Client
  JOINED: 'joined',
  TRANSLATION: 'translation',
  PARTICIPANT_JOINED: 'participant_joined',
  PARTICIPANT_LEFT: 'participant_left',
  ERROR: 'error',
  PONG: 'pong'
};

export function createJoinMessage(roomId, language, name) {
  return {
    type: MessageType.JOIN,
    room_id: roomId,
    language: language,
    name: name
  };
}

export function createAudioMessage(data, timestamp = Date.now()) {
  return {
    type: MessageType.AUDIO,
    data: data, // base64 encoded Opus
    timestamp: timestamp
  };
}

export function createUtteranceEndMessage(timestamp = Date.now()) {
  return {
    type: MessageType.UTTERANCE_END,
    timestamp: timestamp
  };
}

export function createLeaveMessage() {
  return {
    type: MessageType.LEAVE
  };
}

export function createPingMessage() {
  return {
    type: MessageType.PING,
    timestamp: Date.now()
  };
}

export function parseMessage(data) {
  try {
    if (typeof data === 'string') {
      return JSON.parse(data);
    }
    return data;
  } catch (error) {
    console.error('Failed to parse message:', error);
    return null;
  }
}

export function validateJoinedMessage(msg) {
  return (
    msg.type === MessageType.JOINED &&
    typeof msg.room_id === 'string' &&
    typeof msg.participant_id === 'string' &&
    Array.isArray(msg.participants)
  );
}

export function validateTranslationMessage(msg) {
  return (
    msg.type === MessageType.TRANSLATION &&
    typeof msg.speaker_id === 'string' &&
    typeof msg.speaker_name === 'string' &&
    typeof msg.source_lang === 'string' &&
    typeof msg.source_text === 'string' &&
    typeof msg.translations === 'object' &&
    typeof msg.timestamp === 'number'
  );
}

export function validateParticipantMessage(msg) {
  return (
    (msg.type === MessageType.PARTICIPANT_JOINED || msg.type === MessageType.PARTICIPANT_LEFT) &&
    (msg.participant || msg.participant_id)
  );
}

File: server/transport/protocol.py
"""
WebSocket protocol message schemas for Babblefish.
"""
from typing import Dict, List, Optional, Any, Literal
from pydantic import BaseModel, Field


# Client → Server Messages


class JoinMessage(BaseModel):
    """Client joining a room."""

    type: Literal["join"]
    room_id: str
    language: str  # ISO 639-1 code
    name: str
    capabilities: Optional[Dict[str, Any]] = None  # Phase 2/3: client capabilities
    voice_reference: Optional[Dict[str, Any]] = None  # Phase 3: voice reference data


class AudioMessage(BaseModel):
    """Audio chunk from client."""

    type: Literal["audio"]
    data: str  # Base64-encoded Opus audio
    timestamp: int


class UtteranceEndMessage(BaseModel):
    """Client VAD detected end of utterance."""

    type: Literal["utterance_end"]
    timestamp: int


class LeaveMessage(BaseModel):
    """Client leaving room."""

    type: Literal["leave"]


class EnrolMessage(BaseModel):
    """Speaker enrollment audio (Phase 3)."""

    type: Literal["enrol"]
    audio: str  # Base64-encoded PCM audio (15s, 16kHz mono)


class VoiceReferenceMessage(BaseModel):
    """Voice reference from client (Phase 3)."""

    type: Literal["voice_reference"]
    speaker_id: str
    voice_data: str  # Base64-encoded audio
    sample_rate: int
    timestamp: int


# Server → Client Messages


class JoinedMessage(BaseModel):
    """Confirmation that client joined room."""

    type: Literal["joined"] = "joined"
    room_id: str
    participant_id: str
    participants: List[Dict[str, Any]]
    tts_mode: str = "text_only"  # Phase 2/3: assigned TTS mode
    tts_model: Optional[str] = None  # Phase 2/3: assigned TTS model


class TranslationMessage(BaseModel):
    """Translation result broadcast to room."""

    type: Literal["translation"] = "translation"
    speaker_id: str
    speaker_name: str
    source_lang: str
    source_text: str
    translations: Dict[str, str]  # {language: translated_text}
    timestamp: float
    audio: Optional[str] = None  # Phase 2/3: Base64 Opus audio (for server TTS clients)


class ParticipantJoinedMessage(BaseModel):
    """Notification that another participant joined."""

    type: Literal["participant_joined"] = "participant_joined"
    participant: Dict[str, Any]


class ParticipantLeftMessage(BaseModel):
    """Notification that a participant left."""

    type: Literal["participant_left"] = "participant_left"
    participant_id: str


class SpeakerEnrolledMessage(BaseModel):
    """New speaker enrolled (Phase 3)."""

    type: Literal["speaker_enrolled"] = "speaker_enrolled"
    participant_id: str
    name: str
    language: str
    reference_audio: str  # Base64 PCM (~200KB for 15s @ 16kHz)
    reference_text: str  # Whisper transcription


class VoiceReferenceBroadcastMessage(BaseModel):
    """Broadcast voice reference to room participants (Phase 3)."""

    type: Literal["voice_reference_broadcast"] = "voice_reference_broadcast"
    speaker_id: str
    speaker_name: str
    voice_data: str  # Base64-encoded audio
    sample_rate: int
    timestamp: float


class ErrorMessage(BaseModel):
    """Error message to client."""

    type: Literal["error"] = "error"
    code: str
    message: str
